linkerd/app/outbound/src/http/logical/router.rs

```diff
@@ -0,0 +1,290 @@
+///   idle-oriented eviction).
+#[derive(Debug)]
+pub struct NewRouter<N, R, U> {
+    new_backend: N,
+    route_layer: R,
+    _marker: PhantomData<fn(U)>,
+}
+
+#[derive(Clone, Debug)]
+pub struct Router<S>(watch::Receiver<Shared<S>>);
+
+#[derive(Debug)]
+struct Shared<S> {
+    matches: Vec<(RequestMatch, Route)>,
+    routes: HashMap<Route, S>, // TODO(ver) AHashMap?
+}
+
+struct State<T, U, N, S, L, R> {
+    target: T,
+
+    new_backend: N,
+    backends: HashMap<NameAddr, S>,
+
+    route_layer: L,
+    matches: Vec<(RequestMatch, Route)>,
+    routes: HashMap<Route, R>,
+
+    _marker: PhantomData<fn(U)>,
+}
+
+// === impl NewRouter ===
+
+impl<N, R: Clone, U> NewRouter<N, R, U> {
+    pub fn layer(route_layer: R) -> impl layer::Layer<N, Service = Self> + Clone {
+        layer::mk(move |new_backend| Self {
+            new_backend,
+            route_layer: route_layer.clone(),
+            _marker: PhantomData,
+        })
+    }
+}
+
+impl<T, U, N, R, S> NewService<T> for NewRouter<N, R, U>
+where
+    T: Param<profiles::LogicalAddr> + Param<profiles::Receiver>,
+    T: Clone + Send + 'static,
+    U: From<(ConcreteAddr, T)> + Send + 'static,
+    N: NewService<U> + Clone + Send + 'static,
+    N::Service: Clone + Send + Sync + 'static,
+    R: layer::Layer<NewCloneService<Distribute<N::Service>>>,
+    R: Clone + Send + 'static,
+    R::Service: NewService<(Route, T), Service = S>,
+    S: Clone + Send + Sync + 'static,
+{
+    type Service = Router<S>;
+
+    fn new_service(&self, target: T) -> Self::Service {
+        let mut profiles: profiles::Receiver = target.param();
+
+        // Spawn a background task that updates the all routes and backends for the router.
+        let mut state = State::new(target, self.new_backend.clone(), self.route_layer.clone());
+        let (tx, rx) = watch::channel(
+            state
+                .update(&*profiles.borrow_and_update())
+                .expect("initial update must produce a new state"),
+        );
+        tokio::spawn(
+            state
+                .run(profiles, tx)
+                .instrument(tracing::debug_span!("httprouter")),
+        );
+
+        Router(rx)
+    }
+}
+
+impl<N: Clone, R: Clone, U> Clone for NewRouter<N, R, U> {
+    fn clone(&self) -> Self {
+        Self {
+            new_backend: self.new_backend.clone(),
+            route_layer: self.route_layer.clone(),
+            _marker: PhantomData,
+        }
+    }
+}
+
+// === impl Router ===
+
+impl<B, S> Service<http::Request<B>> for Router<S>
+where
+    S: Service<http::Request<B>> + Clone,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = Oneshot<S, http::Request<B>>;
+
+    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), S::Error>> {
+        // TODO(ver) figure out how backpressure should work here. Ideally, we
+        // should only advertise readiness when at least one backend is ready.
```

Comment:
we could maybe wrap the backends in a `SpawnReady` type thing that also has a shared count of how many backends are ready?

alternatively, though, would it make sense for the router itself to shed load if none of its backends are ready?

Reply:
> would it make sense for the router itself to shed load if none of its backends are ready?

In the current setup, yes; generally, no. We could imagine a scenario where there are multiple levels of routing (e.g. applying routes on both logical and concrete services) such that we may want to be able to failover between routers. Or we could have a reason to put a buffer above the router to absorb some amount of backpressure...

But that's not a hard requirement. We could shed load here if it makes sense. But it also seems like it's not really the router's concern.

Reply:
> we could maybe wrap the backends in a SpawnReady type thing that also has a shared count of how many backends are ready?

Yeah, we can probably wrap each backend with a service that spawnready's and publishes the inner service's state (i.e. to an AtomicBool + Notify) so we can peak at recent backend states without having to do any polling here except to be notified of updates.

The caveat is that we'll only really know for backends that have been used and we'll only be able to detect unavailable servers as they are used by the router... but that's probably the right compromise.

None of this seems like a blocker.

Reply:
That sounds a lot like what `failfast::Gate` is already doing; but it would probably be a layering headache to reuse that directly.